rfc/0016-instances-vs-types/0016-instances-vs-types.md

```diff
@@ -0,0 +1,248 @@
+| Status | `DISCUSSION` | <!-- Possible values: DRAFT, DISCUSSION, ACCEPTED, REJECTED -->
+| Responsible | `@rhazn` |
+<!-- 
+  Status Overview:
+  - DRAFT: The RFC is not ready for a review and currently under change. Feel free to already ask for feedback on the structure and contents at this stage.
+  - DISCUSSION: The RFC is open for discussion. Usually, we open a PR to trigger discussions.
+  - ACCEPTED: The RFC was accepted. Create issues to prepare implementation of the RFC.
+  - REJECTED: The RFC was rejected. If another revision emerges, switch to status DRAFT.
+-->
+
+## Summary
+Following the [Design Principles](https://jvalue.github.io/jayvee/docs/dev/design-principles), Jayvee focuses on describing a goal state instead of how to get there. When creating pipelines, programmers use the pipe syntax `->` to connect previously defined `blocks`. Implicitely, the interpreter creates unnamed singleton instances of those blocks and executes them.
+
+This implicit creation can lead to confusion between blocks and block instances. In addition, it makes it impossible to create more than one instance of a block, which will prevent the creation of more complex DAGs.
+
+This RFC introduces minimal syntax changes to allow the creation of block instances, introduces more code structure and clarifies wording.
+
+## Motivation
+It is impossible to create multiple instances of the same block, limiting how complex a DAG can be described.
+
+## Explanation
+The following explanations use blocks as an example but map directly to **Constraints** and **Valuetypes**. That is, constraints and valuetypes can also only be **defined** outside of pipelines and are valuetypes are **instantiated** inside of pipelines (constraints are only instantiated as part of valuetypes).
+
+### Current state
+#### M2 level (focused on blocks only)
+Jayvee currently allows the instantiation of `composite blocktype`s, `builtin blocktype`s and `block`s. `block`s are connected to `blocktype`s by an `oftype` relationship.
+
+```mermaid
+classDiagram
+direction BT
+class BuiltInBlocktype
+class CompositeBlocktype  
+class Block
+class Blocktype { <<abstract>> }
+
+class OftypeRelationship
+
+CompositeBlocktype --|> Blocktype
+BuiltInBlocktype --|> Blocktype
+
+OftypeRelationship "1"-->"1" Block: instance
+OftypeRelationship "*"-->"1" Blocktype: type
+```
+
+#### An example model visualized
+##### The model
+```
+pipeline CarsPipeline {
+	CarsExtractor
+	   	-> CarsTableInterpreter
+		-> CarsLoader;
+
+	block CarsCSVExtractor oftype CSVExtractor {
+		url: "https://gist.githubusercontent.com/noamross/e5d3e859aa0c794be10b/raw/b999fb4425b54c63cab088c0ce2c0d6ce961a563/cars.csv";
+	}
+
+	block CarsTableInterpreter oftype TableInterpreter {
+		header: true;
+		columns: [
+			"name" oftype text,
+			"mpg" oftype decimal,
+		];
+	}
+
+	block CarsLoader oftype SQLiteLoader {
+		table: "Cars";
+		file: "./cars.sqlite";
+	}
+}
+```
+
+##### As diagram including M2 level
```

Comment:
I'd still strongly vote to use the conventions used in similar graphs. We use the names and semantics of `M0`, `M1`, and `M2` from the UML metamodel, so why do we decide to use the exact opposite semantic of the arrows?

 See [Metamodeling on Wikipedia](https://en.wikipedia.org/wiki/Metamodeling) and [an example that is very similar to our case](https://en.wikipedia.org/wiki/Metamodeling#/media/File:M0-m3.png).
![Example](https://upload.wikimedia.org/wikipedia/commons/9/93/M0-m3.png)